### ISSUE TITLE
To diagnose runtime behaviour, developers should be able to enable/read "informational" log messages about workflow execution

### ISSUE THREAD

**Issue by craigfowler** [elsa-workflows/elsa-core#444]
There is some logging already-present in Elsa, but it seems that it is primarily used for error-reporting and little more.  This is a good start but it would be useful if the log could also record informational-level messages.  Whether or not informational messages are actually written to the log output would be down to app configuration.

The motivation for this is so that a developer may see how a workflow executed.  This would help when trying to diagnose unexpected runtime results/behaviour.

## Proposed information to log at "informational" level
This list almost certainly could be expanded, even after this ticket is completed.  It's really offered as a starting point for discussion/expansion.

* The workflow instance Id & Correlation Id (should accompany almost every log message)
* When an activity is executed or resumed, its type, name, Id and the result it returns
    * In the case of an "Outcome" result, the string outcome name(s)
* When a workflow/activity invoker is used to trigger a workflow

## Possible follow-up: Debug-level logging
A logger could also go further than above, and at a more verbose "debug" logging level, it could log the entire serialized state of an activity which is about to be executed.  That would of course create a noisy log, and in production that level of logging would surely be disabled.  But - it would allow developers who are trying to understand a problem to see *exactly what was going on in their workflow* during execution.

Some of the more complex activities (say, receive & write HTTP request) could also benefit from debug logging, recording the raw HTTP requests/responses.

**Comment by sfmskywalker** [comment]
See also #594

**Comment by sfmskywalker** [comment]
Just a quick update on this:

Elsa 2 now has a more detailed execution log store, where activity events such as Executing, Executed and Faulted are logged. Including source file and line number (when using the builder API to define workflows). Outcomes & Outputs are next on the list (#594)

**Comment by craigfowler** [comment]
I'm actually happy that enough of this is done to call it already-resolved.

* Workflow start/stop execution is logged, including the Workflow instance ID
* Activity begin & end execution is logged, including activity type & ID
  * The outcomes are not logged although I don't think its really feasible to do so right now.
  * The `IActivityExecutionResult` interface does not define a `ToString` method or a way of getting a meaningful representation
  * That means that all we have to fall back upon is the name of the actual concrete type of the result object, but this is likely to be insufficient, as many activities return composite/combined results
  * In short, I'm happy to ditch this from the requirements for v2, pending a deeper think about what we really want.  _It's good enough for now_

**Comment by sfmskywalker** [comment]
I agree. Let's close this as done and reiterate as necessary & separately.
